brand/models/commentary.py

```diff
@@ -1,13 +1,50 @@
 from django.core.exceptions import ValidationError
 from django.core.validators import MaxValueValidator, MinValueValidator
 from django.db import models
+from django.utils.safestring import mark_safe
 
 import yaml
+from jsonschema import validate
+from jsonschema.exceptions import ValidationError as YamlValidationError
+from yamlfield.fields import YAMLField
 
 from brand.models import Brand
 from brand.models.embrace_campaign import EmbraceCampaign
 
 
+def validate_feature_override_yaml(feature_yaml) -> None:
+    # Ensure feature_yaml format is of dict/yaml type
+    if isinstance(feature_yaml, dict):
+        feature_override_schema = {
```

Comment:
oh man, I see that this is going. Validation in this way makes sense... However, I don't think that this generic matching is going to work. (I'm sorry!) Instead, you'll have to create a whole validation schema. When the user enters some data, it should navigate to the lowest possible part of that section and validate their result against that. It would be best to put this validation schema in a separate file and import it so that the logic is easier to read.


For example, say that the validation schema is this:
```python
schema = {
    "type": "object",
    "properties": {
        "customers_served": {
            "type": "object",
            "properties": {
                "corporate": {
                    "type": "object",
                    "properties": {
                        "additional_details": {"type": "string"},
                        "offered": {"type": "boolean"},
                    },
                    "required": ["additional_details", "offered"]
                },
                "government": {
                    "type": "object",
                    "properties": {
                        "additional_details": {"type": "string"},
                        "offered": {"type": "boolean"},
                    },
                    "required": ["additional_details", "offered"]
                },
                "nonprofit": {
                    "type": "object",
                    "properties": {
                        "additional_details": {"type": "string"},
                        "offered": {"type": "boolean"},
                    },
                    "required": ["additional_details", "offered"]
                },
                "retail_and_individual": {
                    "type": "object",
                    "properties": {
                        "additional_details": {"type": "string"},
                        "offered": {"type": "boolean"},
                    },
                    "required": ["additional_details", "offered"]
                },
                "smes": {
                    "type": "object",
                    "properties": {
                        "additional_details": {"type": "string"},
                        "offered": {"type": "boolean"},
                    },
                    "required": ["additional_details", "offered"]
                }
            },
        },
        "deposit_products": {"type": ["string", "null"]}  # Allow string or null
    },

}
```

Then the user types in this:
```yaml
customers_served:
  corporate:
    additional_details: blah blah blah
    offered: false
 ```
 
 Now, you validate:
 ```
 validate(RATER_JSON, schema = schema["customers_served"]["corporate"])
 ```
 If the rater adds a separate field (e.g. `government` then you validate again in the same way, and then validate against the higher level `schema["customers_served"]

Reply:
The thing is that your implementation here will allow raters to add really arbitrary keys, which they will do, expecting them to appear on the frontend website magicaclly somehow. We really need to keep them to the allowed keys and be able to support some flexibility in our own schema if necessary later on